src/main/scala/is/hail/io/bgen/BgenWriter.scala

```diff
@@ -0,0 +1,179 @@
+package is.hail.io.bgen
+
+import is.hail.annotations._
+import is.hail.utils._
+import is.hail.variant.{Genotype, Variant, VariantDataset}
+
+object BgenWriter {
+  val ploidy: Byte = 2
+  val phased: Byte = 0
+
+  def shortToBytesLE(bb: ArrayBuilder[Byte], i: Int) {
+    bb += (i & 0xff).toByte
+    bb += ((i >>> 8) & 0xff).toByte
+  }
+
+  def intToBytesLE(bb: ArrayBuilder[Byte], i: Int) {
+    bb += (i & 0xff).toByte
+    bb += ((i >>> 8) & 0xff).toByte
+    bb += ((i >>> 16) & 0xff).toByte
+    bb += ((i >>> 24) & 0xff).toByte
+  }
+
+  def stringToBytesWShortLength(bb: ArrayBuilder[Byte], s: String) {
+    val bytes = s.getBytes
+    shortToBytesLE(bb, bytes.length)
+    bytes.foreach(bb += _)
+  }
+
+  def stringToBytesWIntLength(bb: ArrayBuilder[Byte], s: String) {
+    val bytes = s.getBytes
+    intToBytesLE(bb, bytes.length)
+    bytes.foreach(bb += _)
+  }
+
+  def bitPack(input: Array[Int], nBitsPerProb: Int): Array[Byte] = {
+    val bb = new ArrayBuilder[Byte]
+    bitPack(bb, input, nBitsPerProb)
+    bb.result()
+  }
+
+  def bitPack(bb: ArrayBuilder[Byte], input: Array[Int], nBitsPerProb: Int) {
+    require(nBitsPerProb >= 1 && nBitsPerProb <= 32, "Number of bits per probability must be between 1 and 32 inclusive.")
+
+    val bitMask = (1L << nBitsPerProb) - 1
+    var byteIndex = 0
+    var data = 0L
+    var dataSize = 0
+
+    input.foreach { i =>
+      data |= ((i.toUIntFromRep.toLong & bitMask) << dataSize)
+      dataSize += nBitsPerProb
+
+      while (dataSize >= 8) {
+        bb += (data & 0xffL).toByte
+        data = data >>> 8
+        dataSize -= 8
+        byteIndex += 1
+      }
+    }
+
+    if (dataSize > 0)
+      bb += (data & 0xffL).toByte
+  }
+
+  def resizeProbInts(input: Array[Int], conversionFactor: Double): Array[Int] = {
+    val resized = input.map(i => i.toUIntFromRep * conversionFactor).zipWithIndex.map { case (d, i) => (i, d, math.floor(d)) }
+    val totalFractional = resized.map(d => d._2 - d._3).sum
+
+    resized.sortBy(d => -(d._2 - d._3)).zipWithIndex.map { case (r, i) =>
+      if (i < totalFractional)
+        (r._1, math.ceil(r._2))
+      else
+        (r._1, math.floor(r._2))
+    }.sortBy(_._1).map(_._2.toUInt.intRep)
```

Comment:
OK, this looks right, but the performance is going to be bad.  Basically, don't use any Scala library functions to implement this.  Right now, you're guaranteed input is size 3, right?  Few things:

 - computing totalFractional/F, this is guaranteed to be an integer, so I'd convert it to one.  Since as a floating point value, it might not up exactly to the integer value, round do the nearest integer: `val totalFractional = (sum + 0.5).toInt`.

 - Assert the sum is 2^B-1 at the end.

 - As for the sort, the right way to do this is a function that computes the "sorted index", for each element of an array, it computes the position that element would have in the sorted array.   So if `idx = a.sortedIndex`, then `for (i <- ...) sorted[idx[i]] = a[i]` would be sorted.  I'd allocate the index array on the outside and pass it in so it can be reused.  That way, there are no per-genotype allocations.

 - Finally, to implement `sortedIndex(a, idx)`, input is guaranteed to be length 3, right?  In that case, I might hardcode it.  Otherwise, we should implement it ourself and put it in utils.  This is such a useful function, I don't know why it isn't in the standard library of almost any language.